_posts/2025-04-22-test-classloading-rewrite.adoc

```diff
@@ -0,0 +1,63 @@
+- *Config access from JUnit conditions*. Using a `ConfigProvider` from a custom JUnit condition will https://github.com/quarkusio/quarkus/issues/47081[trigger a `ServiceConfigurationError`]. The workaround is to set the thread context classloader to `this.getClass().getClassLoader()` before reading config, and then set it back afterwards.
+- Increased memory footprint running tests. For suites using multiple profiles and resources, more heap or metaspace may be needed.
+
+=== Things to watch out for
+
+- *Test order change*. A side effect of the rewrite is that the execution order of test can be different. Of course, we all know tests should not depend on execution order if they don't set an order explicitly. However, it's easy to not notice that a test requires a certain order... until the order changes. We discovered some tests in our own suite that were sensitive to the execution order, and other people may make similar discoveries.
+- *Test timing change*. We also discovered that the rewrite exposed some timing issues in tests. Because classloading is frontloaded at the beginning of the test run, rather than between test executions, there's less time for asynchronous operations to finish between tests. For example, there may no longer be time for external state to 'reset' before the next test starts. This might expose some heisenbugs in test suites.
+
+=== Dropped support
+
+- *`@TestProfile` on `@Nested` tests.* Mixing different test profiles and test resources on `@Nested` tests is no longer supported. By definition, every `@TestProfile` must get its own Quarkus application and classloader. Having multiple classloaders execute one test isn't compatible with loading the test with the classloader used to run it.
+- *Version 2.x of the Maven Surefire plugin*. Versions below 3.x of the Maven Surefire plugin will no longer work with `@QuarkusTest`. Version 3 of the Surefire plugin was released in 2023, so version 2 is now rather old.
+
+== Next steps
+
+The main work of the test classloading rewrite has been delivered in 3.22, and has unlocked a bunch of possible improvements.
+Some test defects weren't directly fixed by the main change, but the architecture is now in place to enable a fix.
+More excitingly, test-related extensions, like the Pact extensions, can now add new features to reduce test boilerplate.
+
+As always, if you spot issues or oddities, please let us know on https://quarkusio.zulipchat.com/[zulip] or https://github.com/quarkusio/quarkus/issues[raise an issue].
+
```

Comment:
I would mention the working group somewhere :-)

Reply:
Yes, good point! I had every intention of mentioning it, and even pasted the link into a draft, but lost it. Will do.